src/QirRuntime/test/QIR-tracer/tracer-target.qs

```diff
@@ -0,0 +1,297 @@
+    is Adj + Ctl {
+        body  (...) { Phys.single_qubit_op(23, 1, qb); }
+        adjoint (...) { Phys.single_qubit_op(24, 1, qb); }
+        controlled (ctls, ...) { Phys.single_qubit_op_ctl(25, 1, ctls, qb); }
+        controlled adjoint (ctls, ...) { Phys.single_qubit_op_ctl(25, 1, ctls, qb); }
+    }
+
+    @Inline()
+    operation Mz(qb : Qubit) : Result {
+        body  (...) { return Phys.single_qubit_measure(100, 1, qb); }
+    }
+
+    @Inline()
+    operation Mx(qb : Qubit) : Result {
+        body  (...) { return Phys.single_qubit_measure(101, 1, qb); }
+    }
+
+    @Inline()
+    operation M(qb : Qubit) : Result {
+        body  (...) { return Mz(qb); }
+    }
+
+    @Inline()
+    operation Mzz(qubits : Qubit[]) : Result {
+        body  (...) { return Phys.multi_qubit_measure(102, 1, qubits); }
+    }
+
+    @Inline()
+    operation Mxz(qubits : Qubit[]) : Result {
+        body  (...) { return Phys.multi_qubit_measure(103, 1, qubits); }
+    }
+
+    @Inline()
+    operation Mzx(qubits : Qubit[]) : Result {
+        body  (...) { return Phys.multi_qubit_measure(104, 1, qubits); }
+    }
+
+    @Inline()
+    operation Mxx(qubits : Qubit[]) : Result {
+        body  (...) { return Phys.multi_qubit_measure(105, 1, qubits); }
+    }
+
+    @Inline()
+    operation Measure(paulis : Pauli[], qubits : Qubit[]) : Result {
+        body  (...)
+        {
+            mutable res = One;
+            mutable haveY = false;
+            // Measurements that involve PauliY or PauliI
+            for i in 0..Length(paulis)-1
+            {
+                if (paulis[i] == PauliY or paulis[i] == PauliI)
+                {
+                    set haveY = true;
+                }
+            }
+            if (haveY) { set res = Phys.multi_qubit_measure(106, 1, qubits); }
+
+            // More than two qubits (but no PauliY or PauliI)
+            elif (Length(paulis) > 2) { set res = Phys.multi_qubit_measure(107, 1, qubits); }
+
+            // Single qubit measurement -- differentiate between Mx and Mz
+            elif (Length(paulis) == 1)
+            {
+                if (paulis[0] == PauliX) { set res = Mx(qubits[0]); }
+                else { set res = Mz(qubits[0]); }
+            }
+
+            // Specialize for two-qubit measurements: Mxx, Mxz, Mzx, Mzz
+            elif (paulis[0] == PauliX and paulis[1] == PauliX) { set res = Phys.multi_qubit_measure(108, 1, qubits); }
+            elif (paulis[0] == PauliX and paulis[1] == PauliZ) { set res = Phys.multi_qubit_measure(109, 1, qubits); }
+            elif (paulis[0] == PauliZ and paulis[1] == PauliX) { set res = Phys.multi_qubit_measure(110, 1, qubits); }
+            elif (paulis[0] == PauliZ and paulis[1] == PauliZ) { set res = Phys.multi_qubit_measure(111, 1, qubits); }
+
+            //shouldn't get here
+            return res;
+        }
+    }
+
+
+    @TargetInstruction("inject_global_barrier")
+    operation Barrier(id : Int, duration : Int) : Unit
```

Comment:
Similarly here, this looks like a proposal for a new Q# intrinsic operation? If so, would be willing to write an issue on QuantumLibraries so that we can get it scheduled for the next API review? Thanks!

Reply:
@bettinaheim, I suspect that eventually we'll need barriers to be in the public interface but I'm not sure of the timeline. For now, I've moved it into the Tracer namespace to allow for experimentation and testing but should we kick off the motions to add it to the libraries as well?

Reply:
I think that would make sense, yeah. No harm in getting the Q# API reviewed in advance so that the approval is ready and in place for whatever timelines work on the C&R side.